sound/soc/sof/topology.c

```diff
@@ -1257,6 +1279,84 @@ static int sof_widget_parse_tokens(struct snd_soc_component *scomp, struct snd_s
 	return ret;
 }
 
+static int sof_alloc_pin_binding(struct snd_sof_widget *swidget,
+				 bool for_sink)
+{
+	int ret = 0;
+
+	char **pin_binding;
+	int num_pins;
+	int i, j;
+
+	if (for_sink) {
+		pin_binding = swidget->sink_pin_binding;
+		num_pins = swidget->num_sink_pins;
+	} else {
+		pin_binding = swidget->src_pin_binding;
+		num_pins = swidget->num_source_pins;
+	}
+
+	for (i = 0; i < num_pins; i++) {
+		pin_binding[i] = kzalloc(SNDRV_CTL_ELEM_ID_NAME_MAXLEN,
+					 GFP_KERNEL);
```

Comment:
@aiChaoSONG I think we can use kstrdup to allocate and copy string in get_token_string. Please see [pin_binding.txt](https://github.com/thesofproject/linux/files/9258243/pin_binding.txt) for reference.